nimble/host/services/audio/bass/src/ble_audio_svc_bass.c

```diff
@@ -0,0 +1,915 @@
+    }, {
+        .op_code = BLE_AUDIO_SVC_BASS_REMOVE_SOURCE,
+        .expected_len = 1,
+        .handler_cb = ble_svc_audio_bass_remove_source
+    }
+};
+
+static int
+ble_svc_audio_bass_access(uint16_t conn_handle, uint16_t attr_handle,
+                          struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+static int
+ble_svc_audio_bass_ctrl_point_write_access(struct ble_gatt_access_ctxt *ctxt);
+static int
+ble_svc_audio_bass_rcv_state_read_access(struct ble_gatt_access_ctxt *ctxt, void *arg);
+
+static struct ble_gatt_chr_def ble_svc_audio_bass_chrs[MYNEWT_VAL(BLE_SVC_AUDIO_BASS_RECEIVE_STATE_MAX) + 2];
+
+static const struct ble_gatt_svc_def ble_svc_audio_bass_defs[MYNEWT_VAL(BLE_SVC_AUDIO_BASS_RECEIVE_STATE_MAX) + 2] = {
+    { /*** Service: Published Audio Capabilities Service (bass) */
+        .type = BLE_GATT_SVC_TYPE_PRIMARY,
+        .uuid = BLE_UUID16_DECLARE(BLE_SVC_AUDIO_BASS_UUID16),
+        .characteristics = ble_svc_audio_bass_chrs,
+    },
+    {
+        0, /* No more services. */
+    },
+};
+
+static int
+ble_svc_audio_bass_access(uint16_t conn_handle, uint16_t attr_handle,
+                          struct ble_gatt_access_ctxt *ctxt, void *arg)
+{
+    uint16_t uuid16 = ble_uuid_u16(ctxt->chr->uuid);
+    int rc;
+
+    switch (uuid16) {
+    case BLE_SVC_AUDIO_BASS_CHR_UUID16_BAS_CONTROL_POINT:
+        if (ctxt->op == BLE_GATT_ACCESS_OP_WRITE_CHR) {
+            rc = ble_svc_audio_bass_ctrl_point_write_access(ctxt);
+        } else {
+            assert(0);
+        }
+        return rc;
+    case BLE_SVC_AUDIO_BASS_CHR_UUID16_BROADCAST_RECEIVE_STATE:
+        if (ctxt->op == BLE_GATT_ACCESS_OP_READ_CHR) {
+            rc = ble_svc_audio_bass_rcv_state_read_access(ctxt, arg);
+        } else {
+            assert(0);
+        }
+        return rc;
+    default:
+        assert(0);
+    }
+}
+
+static uint8_t
+ble_svc_audio_bass_get_new_source_id()
+{
+    struct ble_svc_audio_bass_receiver_state *rcv_state;
+    struct ble_svc_audio_bass_receiver_state *next_rcv_state;
+    uint8_t free_source_id;
+    uint8_t highest_source_id = 0;
+    uint8_t lowest_source_id = BLE_SVC_AUDIO_BASS_RECEIVE_STATE_SRC_ID_NONE;
+
+    int i;
+
+    /* First characteristic is BLE_SVC_AUDIO_BASS_CHR_UUID16_BAS_CONTROL_POINT, iterate from 2nd */
+    for (i = 1; i <= MYNEWT_VAL(BLE_SVC_AUDIO_BASS_RECEIVE_STATE_MAX); i++) {
+        rcv_state = ble_svc_audio_bass_chrs[i].arg;
+        if (highest_source_id < rcv_state->source_id &&
+            rcv_state->source_id != BLE_SVC_AUDIO_BASS_RECEIVE_STATE_SRC_ID_NONE)
+        {
+            highest_source_id = rcv_state->source_id;
+        }
+
+        if (lowest_source_id > rcv_state->source_id) {
+            lowest_source_id = rcv_state->source_id;
+        }
+    }
+
+    if ((highest_source_id + 1) < BLE_SVC_AUDIO_BASS_RECEIVE_STATE_SRC_ID_NONE) {
+        return highest_source_id + 1;
+    }
+
+    return min(lowest_source_id + 1, 0);
```

Comment:
This does not seem right to me. Consider having 3 Receiver states with sources added, let say 1, 2, 3.
1. Client removes source 3.
2. Client adds new source.
3. The source ID assigned will be 3.

I think the better approach would be to assign source ID 4. It's because it will be clear for e.g. a second Broadcast Assistent that will connect and read the Receiver States that this Source is new, and the old one has been removed.

Consider having a static variable with last or first free source ID, so that you won't need to loop over the registered sources.